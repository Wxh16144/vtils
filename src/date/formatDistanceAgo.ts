import { formatDistanceToNowStrict } from 'date-fns/esm'
import { zhCN } from 'date-fns/esm/locale'

/**
 * 将时间转化为 `xxx前` 表示。
 *
 * @param date 时间
 */
export function formatDistanceAgo(date: number | Date): string {
  const distance = formatDistanceToNowStrict(date, {
    locale: zhCN,
  })
  return `${distance.replace(/\s+/g, '')}前`
}
