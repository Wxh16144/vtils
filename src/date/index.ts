/**
 * 日期时间工具库。基于 {@link https://date-fns.org/v2.14.0/docs/Getting-Started | date-fns}。
 *
 * @packageDocumentation
 */

/* istanbul ignore file */

// prettier-ignore
export {
  // 手动导出以解决 @microsoft/api-extractor 的问题
  // @index(['../../node_modules/date-fns/esm/*', '!**/{_*,locale,constants,fp}'], (f, _) => `${f.name.replace(/-/g, '')},`, { onlyDirectories: true })
  add,
  addBusinessDays,
  addDays,
  addHours,
  addISOWeekYears,
  addMilliseconds,
  addMinutes,
  addMonths,
  addQuarters,
  addSeconds,
  addWeeks,
  addYears,
  areIntervalsOverlapping,
  closestIndexTo,
  closestTo,
  compareAsc,
  compareDesc,
  differenceInBusinessDays,
  differenceInCalendarDays,
  differenceInCalendarISOWeeks,
  differenceInCalendarISOWeekYears,
  differenceInCalendarMonths,
  differenceInCalendarQuarters,
  differenceInCalendarWeeks,
  differenceInCalendarYears,
  differenceInDays,
  differenceInHours,
  differenceInISOWeekYears,
  differenceInMilliseconds,
  differenceInMinutes,
  differenceInMonths,
  differenceInQuarters,
  differenceInSeconds,
  differenceInWeeks,
  differenceInYears,
  eachDayOfInterval,
  eachHourOfInterval,
  eachMonthOfInterval,
  eachQuarterOfInterval,
  eachWeekendOfInterval,
  eachWeekendOfMonth,
  eachWeekendOfYear,
  eachWeekOfInterval,
  eachYearOfInterval,
  endOfDay,
  endOfDecade,
  endOfHour,
  endOfISOWeek,
  endOfISOWeekYear,
  endOfMinute,
  endOfMonth,
  endOfQuarter,
  endOfSecond,
  endOfToday,
  endOfTomorrow,
  endOfWeek,
  endOfYear,
  endOfYesterday,
  format,
  formatDistance,
  formatDistanceStrict,
  formatDistanceToNow,
  formatDistanceToNowStrict,
  formatDuration,
  formatISO,
  formatISO9075,
  formatISODuration,
  formatRelative,
  formatRFC3339,
  formatRFC7231,
  fromUnixTime,
  getDate,
  getDay,
  getDayOfYear,
  getDaysInMonth,
  getDaysInYear,
  getDecade,
  getHours,
  getISODay,
  getISOWeek,
  getISOWeeksInYear,
  getISOWeekYear,
  getMilliseconds,
  getMinutes,
  getMonth,
  getOverlappingDaysInIntervals,
  getQuarter,
  getSeconds,
  getTime,
  getUnixTime,
  getWeek,
  getWeekOfMonth,
  getWeeksInMonth,
  getWeekYear,
  getYear,
  intervalToDuration,
  isAfter,
  isBefore,
  isDate,
  isEqual,
  isExists,
  isFirstDayOfMonth,
  isFriday,
  isFuture,
  isLastDayOfMonth,
  isLeapYear,
  isMatch,
  isMonday,
  isPast,
  isSameDay,
  isSameHour,
  isSameISOWeek,
  isSameISOWeekYear,
  isSameMinute,
  isSameMonth,
  isSameQuarter,
  isSameSecond,
  isSameWeek,
  isSameYear,
  isSaturday,
  isSunday,
  isThisHour,
  isThisISOWeek,
  isThisMinute,
  isThisMonth,
  isThisQuarter,
  isThisSecond,
  isThisWeek,
  isThisYear,
  isThursday,
  isToday,
  isTomorrow,
  isTuesday,
  isValid,
  isWednesday,
  isWeekend,
  isWithinInterval,
  isYesterday,
  lastDayOfDecade,
  lastDayOfISOWeek,
  lastDayOfISOWeekYear,
  lastDayOfMonth,
  lastDayOfQuarter,
  lastDayOfWeek,
  lastDayOfYear,
  lightFormat,
  max,
  milliseconds,
  min,
  parse,
  parseISO,
  parseJSON,
  roundToNearestMinutes,
  set,
  setDate,
  setDay,
  setDayOfYear,
  setHours,
  setISODay,
  setISOWeek,
  setISOWeekYear,
  setMilliseconds,
  setMinutes,
  setMonth,
  setQuarter,
  setSeconds,
  setWeek,
  setWeekYear,
  setYear,
  startOfDay,
  startOfDecade,
  startOfHour,
  startOfISOWeek,
  startOfISOWeekYear,
  startOfMinute,
  startOfMonth,
  startOfQuarter,
  startOfSecond,
  startOfToday,
  startOfTomorrow,
  startOfWeek,
  startOfWeekYear,
  startOfYear,
  startOfYesterday,
  sub,
  subBusinessDays,
  subDays,
  subHours,
  subISOWeekYears,
  subMilliseconds,
  subMinutes,
  subMonths,
  subQuarters,
  subSeconds,
  subWeeks,
  subYears,
  toDate,
  // @endindex
} from 'date-fns/esm'

// prettier-ignore
export {
  // 手动导出以解决 @microsoft/api-extractor 的问题
  // ref: https://github.com/date-fns/date-fns/blob/master/scripts/_lib/listLocales.js#L13
  // ref: https://github.com/date-fns/date-fns/blob/master/outdatedLocales.json
  // @index(['../../node_modules/date-fns/esm/locale/*', '!**/{_*,ar,fil,fr-CH,nl-BE}'], (f, _) => `${f.name.replace(/-/g, '')},`, { onlyDirectories: true })
  af,
  arDZ,
  arMA,
  arSA,
  az,
  be,
  bg,
  bn,
  ca,
  cs,
  cy,
  da,
  de,
  el,
  enAU,
  enCA,
  enGB,
  enIN,
  enNZ,
  enUS,
  enZA,
  eo,
  es,
  et,
  eu,
  faIR,
  fi,
  fr,
  frCA,
  gd,
  gl,
  gu,
  he,
  hi,
  hr,
  hu,
  hy,
  id,
  is,
  it,
  ja,
  ka,
  kk,
  kn,
  ko,
  lb,
  lt,
  lv,
  mk,
  ms,
  mt,
  nb,
  nl,
  nn,
  pl,
  pt,
  ptBR,
  ro,
  ru,
  sk,
  sl,
  sr,
  srLatn,
  sv,
  ta,
  te,
  th,
  tr,
  ug,
  uk,
  uz,
  vi,
  zhCN,
  zhTW,
  // @endindex
} from 'date-fns/esm/locale'

// @index(['./**/*.ts', '!./**/*.test.*'], f => `export * from '${f.path}'`)
export * from './anyToDate'
export * from './formatDate'
export * from './formatDistanceAgo'
export * from './intervalToRestrictiveDuration'
export * from './numeralDayToChineseDay'
// @endindex
