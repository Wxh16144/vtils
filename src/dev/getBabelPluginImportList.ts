import * as dateFns from 'date-fns'
import * as dateFnsLocale from 'date-fns/locale'
import * as lodash from 'lodash-uni'
import * as path from 'path'

export function getBabelPluginImportList() {
  // @ts-ignore
  const resolve = require.resolve
  const vtilsPath = path.dirname(resolve('vtils/package.json'))

  return [
    // utils
    [
      resolve('babel-plugin-import'),
      {
        libraryName: 'vtils',
        camel2DashComponentName: false,
        customName: (name: string) => {
          if (name in lodash) {
            return `lodash-es/${name}`
          }
          return `vtils/utils/${name}`
        },
      },
      'vtils.utils.0',
    ],
    [
      resolve('babel-plugin-import'),
      {
        libraryName: 'vtils/utils',
        camel2DashComponentName: false,
        customName: (name: string) => {
          if (name in lodash) {
            return `lodash-es/${name}`
          }
          return `vtils/utils/${name}`
        },
      },
      'vtils.utils.1',
    ],
    [
      resolve('babel-plugin-import'),
      {
        libraryName: 'lodash-uni',
        camel2DashComponentName: false,
        customName: (name: string) => `lodash-es/${name}`,
      },
      'vtils.utils.2',
    ],
    [
      resolve('babel-plugin-import'),
      {
        libraryName: '../utils',
        camel2DashComponentName: false,
        customName: (name: string, file: { opts: { filename: string } }) => {
          if (!file.opts.filename.startsWith(vtilsPath)) {
            return '../utils'
          }
          if (name in lodash) {
            return `lodash-es/${name}`
          }
          return `../utils/${name}`
        },
      },
      'vtils.utils.3',
    ],

    // mp
    [
      resolve('babel-plugin-import'),
      {
        libraryName: 'vtils/mp',
        camel2DashComponentName: false,
        customName: (name: string) => `vtils/mp/${name}`,
      },
      'vtils.mp.0',
    ],

    // react
    [
      resolve('babel-plugin-import'),
      {
        libraryName: 'vtils/react',
        camel2DashComponentName: false,
        customName: (name: string) => `vtils/react/${name}`,
      },
      'vtils.react.0',
    ],

    // date
    [
      resolve('babel-plugin-import'),
      {
        libraryName: 'vtils/date',
        camel2DashComponentName: false,
        customName: (name: string) => {
          if (name in dateFns) {
            return `date-fns/esm/${name}`
          }
          if (name in dateFnsLocale) {
            return `date-fns/esm/locale/${name}`
          }
          return `vtils/date/${name}`
        },
      },
      'vtils.date.0',
    ],
    [
      resolve('babel-plugin-import'),
      {
        libraryName: 'date-fns/esm',
        camel2DashComponentName: false,
        customName: (name: string) => `date-fns/esm/${name}`,
      },
      'vtils.date.1',
    ],
    [
      resolve('babel-plugin-import'),
      {
        libraryName: 'date-fns/esm/locale',
        camel2DashComponentName: false,
        customName: (name: string) => `date-fns/esm/locale/${name}`,
      },
      'vtils.date.2',
    ],
  ]
}
