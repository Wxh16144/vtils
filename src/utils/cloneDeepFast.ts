// @ts-nocheck
// Copyright 2019 "David Mark Clements <david.mark.clements@gmail.com>"
// MIT Licensed
// https://github.com/davidmarkclements/rfdc
// Modified by Jay Fong
/**
 * 深克隆快速版。
 *
 * @param value 要克隆的值
 * @returns 返回克隆后的值
 * @example
 * ```typescript
 * cloneDeepFast({ x: [1] })
 * // => { x: [1] }
 * ```
 */
export function cloneDeepFast<T>(value: T): T {
  if (typeof value !== 'object' || value === null) return value
  if (value instanceof Date) return new Date(value)
  if (Array.isArray(value)) return cloneArray(value)
  const o2 = {}
  for (const k in value) {
    if (Object.hasOwnProperty.call(value, k) === false) continue
    const cur = value[k]
    if (typeof cur !== 'object' || cur === null) {
      o2[k] = cur
    } else if (cur instanceof Date) {
      o2[k] = new Date(cur)
    } else {
      o2[k] = cloneDeepFast(cur)
    }
  }
  return o2
}

function cloneArray(a) {
  const keys = Object.keys(a)
  const a2 = new Array(keys.length)
  for (let i = 0; i < keys.length; i++) {
    const k = keys[i]
    const cur = a[k]
    if (typeof cur !== 'object' || cur === null) {
      a2[k] = cur
    } else if (cur instanceof Date) {
      a2[k] = new Date(cur)
    } else {
      a2[k] = cloneDeepFast(cur)
    }
  }
  return a2
}
