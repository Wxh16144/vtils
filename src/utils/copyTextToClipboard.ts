/**
 * 复制文本到剪切板。
 *
 * @param text 要复制的文本
 * @returns 返回是否复制成功
 */
export function copyTextToClipboard(text: string): boolean {
  // https://github.com/sindresorhus/copy-text-to-clipboard/blob/master/index.js
  const element = document.createElement('textarea')

  element.value = text

  // Prevent keyboard from showing on mobile
  element.setAttribute('readonly', '')

  // @ts-ignore
  element.style.contain = 'strict'
  element.style.position = 'absolute'
  element.style.left = '-9999px'
  element.style.fontSize = '12pt' // Prevent zooming on iOS

  const selection = document.getSelection()!
  let originalRange: Range | undefined
  if (selection.rangeCount > 0) {
    originalRange = selection.getRangeAt(0)
  }

  document.body.appendChild(element)
  element.select()

  // Explicit selection workaround for iOS
  element.selectionStart = 0
  element.selectionEnd = text.length

  let isSuccess = false
  try {
    isSuccess = document.execCommand('copy')
  } catch (_) {}

  document.body.removeChild(element)

  if (originalRange) {
    selection.removeAllRanges()
    selection.addRange(originalRange)
  }

  return isSuccess
}
