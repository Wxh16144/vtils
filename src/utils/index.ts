/**
 * 基础工具库。基于 {@link https://lodash.com/docs/4.17.15 | Lodash}。
 *
 * @packageDocumentation
 */

/* istanbul ignore file */

export * from 'lodash-uni'

// @index(['./**/*.ts', '!./**/*.test.*'], f => `export * from '${f.path}'`)
export * from './base64'
export * from './bindEvent'
export * from './chooseFile'
export * from './constantCase'
export * from './copyTextToClipboard'
export * from './createSubmit'
export * from './createUrlQueryString'
export * from './dedent'
export * from './EventBus'
export * from './formatBytes'
export * from './formatNumber'
export * from './getEnvironment'
export * from './getWechatPublicAccountQrcodeUrl'
export * from './inAndroid'
export * from './inBrowser'
export * from './inDeno'
export * from './indent'
export * from './inIOS'
export * from './inMiniProgram'
export * from './inNodeJS'
export * from './inTaro'
export * from './inWechatWebView'
export * from './isBlobUrl'
export * from './isChineseIDCardNumber'
export * from './isDataUrl'
export * from './isNumeric'
export * from './isPossibleChineseMobilePhoneNumber'
export * from './isPromiseLike'
export * from './isUrl'
export * from './keysStrict'
export * from './loadCss'
export * from './loadResource'
export * from './md5'
export * from './move'
export * from './omitStrict'
export * from './onceMeanwhile'
export * from './parseDataUrl'
export * from './parseUrlQueryString'
export * from './pascalCase'
export * from './pickStrict'
export * from './placeKitten'
export * from './readFile'
export * from './RichUrl'
export * from './run'
export * from './swap'
export * from './traverse'
export * from './TreeData'
export * from './wait'
export * from './Wechat'
// @endindex
