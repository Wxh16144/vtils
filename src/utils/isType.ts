/* eslint-disable @typescript-eslint/no-unused-vars */
/**
 * 对 `value is T` 的封装。
 */
export function isType<T>(value: any): value is T {
  return true
}
