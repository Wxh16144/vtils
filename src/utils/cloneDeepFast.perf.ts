import { cloneDeep, range } from 'lodash-uni'
import { cloneDeepFast } from './cloneDeepFast'
import { runBenchmark } from '../dev'

const data = range(0, 100).map(i => ({
  i,
  c: range(0, 10).map(i => ({ i: [i] })),
}))

runBenchmark({
  ['lodash.cloneDeep']() {
    cloneDeep(data)
  },
  ['vtils.cloneDeepFast!fastest']() {
    cloneDeepFast(data)
  },
})
