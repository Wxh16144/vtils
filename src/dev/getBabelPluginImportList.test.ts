import * as babel from '@babel/core'
import { dedent } from '../utils'
import { getBabelConfig } from 'haoma'
import { getBabelPluginImportList } from './getBabelPluginImportList'

describe('getBabelPluginImportList', () => {
  test('转换正常', async () => {
    const code = dedent`
      import { sample, sampleIndex } from 'vtils'
      import { wait, sampleBy } from '../utils'
      import { uniq } from 'lodash-uni'
      import { anyToDate, getDate, zhCN } from 'vtils/date'
      import { getDay } from 'date-fns/esm'
      import { de } from 'date-fns/esm/locale'
      import { getTopBarInfo } from 'vtils/mp'

      console.log(sample, sampleIndex, wait, sampleBy, anyToDate, getDate, zhCN, getDay, de, uniq, getTopBarInfo)
    `
    const res = await babel.transformAsync(
      code,
      getBabelConfig({
        filename: __filename,
        target: 'browser',
        module: 'cjs',
        plugins: getBabelPluginImportList(),
      }),
    )
    expect(res!.code!).toMatchSnapshot()
  })
})
