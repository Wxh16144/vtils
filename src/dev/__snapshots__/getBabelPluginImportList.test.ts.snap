// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBabelPluginImportList 转换正常 1`] = `
"\\"use strict\\";

var _interopRequireDefault = require(\\"@babel/runtime/helpers/interopRequireDefault\\");

var _de2 = _interopRequireDefault(require(\\"date-fns/esm/locale/de\\"));

var _getDay2 = _interopRequireDefault(require(\\"date-fns/esm/getDay\\"));

var _zhCN2 = _interopRequireDefault(require(\\"date-fns/esm/locale/zhCN\\"));

var _getDate2 = _interopRequireDefault(require(\\"date-fns/esm/getDate\\"));

var _anyToDate2 = _interopRequireDefault(require(\\"vtils/date/anyToDate\\"));

var _getTopBarInfo2 = _interopRequireDefault(require(\\"vtils/mp/getTopBarInfo\\"));

var _utils = _interopRequireDefault(require(\\"../utils\\"));

var _uniq2 = _interopRequireDefault(require(\\"lodash-es/uniq\\"));

var _sampleIndex2 = _interopRequireDefault(require(\\"vtils/utils/sampleIndex\\"));

var _sample2 = _interopRequireDefault(require(\\"lodash-es/sample\\"));

console.log(_sample2.default, _sampleIndex2.default, _utils.default, _utils.default, _anyToDate2.default, _getDate2.default, _zhCN2.default, _getDay2.default, _de2.default, _uniq2.default, _getTopBarInfo2.default);"
`;
